prow/config/cache.go

```diff
@@ -0,0 +1,388 @@
+	}
+	keyParts.Identifier = identifier
+
+	// Append "baseSHA" string information.
+	baseSHA, err := baseSHAGetter()
+	if err != nil {
+		return CacheKeyParts{}, fmt.Errorf("failed to get baseSHA: %v", err)
+	}
+	keyParts.BaseSHA = baseSHA
+
+	// Append "headSHAs" string information.
+	var headSHAs []string
+	for _, headSHAGetter := range headSHAGetters {
+		headSHA, err := headSHAGetter()
+		if err != nil {
+			return CacheKeyParts{}, fmt.Errorf("failed to get headRef: %v", err)
+		}
+		headSHAs = append(headSHAs, headSHA)
+	}
+
+	keyParts.HeadSHAs = headSHAs
+
+	return keyParts, nil
+}
+
+// MakeCacheKey converts a CacheKeyParts object into a JSON string (to be used
+// as a CacheKey).
+func MakeCacheKey(kp CacheKeyParts) (CacheKey, error) {
+	data, err := json.Marshal(kp)
+
+	return CacheKey(data), err
+}
+
+// GetPresubmitsFromCache is like GetPresubmits, but attempts to use a cache
+// lookup to get the *ProwYAML value (cache hit), instead of computing it from
+// scratch (cache miss).
+func (c *Config) GetPresubmitsFromCache(gc git.ClientFactory, identifier string, baseSHAGetter RefGetter, headSHAGetters ...RefGetter) ([]Presubmit, error) {
+
+	prowYAML, err := GetProwYAMLFromCache(c.ProwYAMLCache, c.getProwYAMLNoDefault, gc, identifier, baseSHAGetter, headSHAGetters...)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := DefaultAndValidateProwYAML(c, prowYAML, identifier); err != nil {
+		return nil, err
+	}
+
+	return append(c.GetPresubmitsStatic(identifier), prowYAML.Presubmits...), nil
+}
+
+// GetPostsubmitsFromCache is like GetPostsubmits, but attempts to use a cache
+// lookup to get the *ProwYAML value (cache hit), instead of computing it from
+// scratch (cache miss).
+func (c *Config) GetPostsubmitsFromCache(gc git.ClientFactory, identifier string, baseSHAGetter RefGetter, headSHAGetters ...RefGetter) ([]Postsubmit, error) {
+
+	prowYAML, err := GetProwYAMLFromCache(c.ProwYAMLCache, c.getProwYAMLNoDefault, gc, identifier, baseSHAGetter, headSHAGetters...)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := DefaultAndValidateProwYAML(c, prowYAML, identifier); err != nil {
+		return nil, err
+	}
+
+	return append(c.GetPostsubmitsStatic(identifier), prowYAML.Postsubmits...), nil
+}
+
+// GetProwYAMLFromCache uses ProwYAMLCache to first try to perform a lookup of
+// previously-calculated ProwYAML objects. The 'valConstructor' function is
+// taken as an argument to make it easier to test this function. This way, unit
+// tests can just provide its own function for constructing a ProwYAML object
+// (instead of needing to create an actual Git repo, etc.).
+func GetProwYAMLFromCache(
+	prowYAMLCache *ProwYAMLCache,
+	valConstructorHelper func(git.ClientFactory, string, RefGetter, ...RefGetter) (*ProwYAML, error),
+	gc git.ClientFactory,
+	identifier string,
+	baseSHAGetter RefGetter,
+	headSHAGetters ...RefGetter) (*ProwYAML, error) {
+
+	keyConstructor := func() (CacheKey, error) {
```

Comment:
Do we this to be wrapped in a function? I think we can just calculate the cache key and pass it rather than a key constructor function.

Reply:
Done